### ISSUE TITLE
How to use BatchNorm in the correct way?

### ISSUE THREAD

**Issue by ufukcbicici** [dmlc/mxnet#1635]
Hello,

I am trying to use Batch Normalization in the project I am working on. Since BatchNorm has different use cases for training and validation, I wanted to be sure what is the proper way to use the operator. When I include BatchNorm in my network, I get two argument buffers (gamma and beta) which correspond to the coefficient and bias terms of the BN transformation, as stated in the original paper http://arxiv.org/pdf/1502.03167v1.pdf. And I have two auxiliary buffers for the moving average and moving variance, which are to be used in the validation phase. My question is, what I need to do use these both correctly? For example, the inputs must be transformed with an affine function where the coefficient and bias terms are calculated with learned gamma-beta and moving average-moving variance parameters, in the validation phase. When I execute the 'forward' method of the executor, produced out of this network, there is an argument 'is_train', set to true-false according to the current calling purpose. Does this flag somehow informs the BatchNorm operator of the correct use case? Or do I need to do calculate the transformation parameters by hand, for example?

Thanks in advance.

**Comment by ascust** [comment]
For training, which means is_train==True, the mean is calculated by using the batch statistics, and so is the var. After this the data is subtracted by the mean and divided by the var. This happens in forward phase. In backward, apart from back propagate the error, the moving_mean and moving_var get updated  by using momentum. If your momentum is small, the moving_mean and moving_var can change a little bit fast. For inference, the batch_mean and batch_var are no longer calculated. Instead, it uses moving_mean and moving_var as kind of global statistics.

**Comment by ufukcbicici** [comment]
Thanks for the answer, it cleared a lot of things for me. I have another question, in the following, the formula for calculating the transform in the validation phase is given:

![batchnorm](https://cloud.githubusercontent.com/assets/17207330/13786541/5aa44e50-ea95-11e5-96bc-8899c64c3d5e.png)

So, the learned gamma and beta is combined with the moving_mean and moving_var in order to final transform. Does the BatchNorm operator calculates it in that way as well: Divide the learned gamma by the square root of the moving variance plus epsilon, scale the input x with that value and add [beta minus (gamma times moving_mean) divided by (square root of the moving variance plus epsilon)]  Do we have the same transformation, when is_train==False in the forward pass?

**Comment by ascust** [comment]
I think so, in the original C++ file, it is like:
```
Assign(out, req[batchnorm::kOut], broadcast<1>(slope /
                                          F<mshadow_op::square_root>(moving_var + param_.eps),
                                          data.shape_) * data +
             broadcast<1>(bias - (slope * moving_mean) /
                          F<mshadow_op::square_root>(moving_var + param_.eps), data.shape_));

```
ignore the irrelevant signs, and you can see it is the same form as in the original paper.
